### PR TITLE
Support `fourmolu ^>= 0.7`

### PR DESCRIPTION
**Pull request by parsonsmatt**
This PR adds support for `fourmolu-0.7`

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2944"></a>

### CODE REVIEW

File: plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal [hunk 1]
```diff
@@ -9,21 +9,27 @@ license:            Apache-2.0
 license-file:       LICENSE
 author:             The Haskell IDE Team
 copyright:          The Haskell IDE Team
+homepage:           https://github.com/haskell/haskell-language-server
+bug-reports:        https://github.com/haskell/haskell-language-server/issues
```
Comment (parsonsmatt) [inline plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal:13]: It took me a minute to figure out the Git repository for this package, so I added these fields to the `.cabal` file. Technically an unrelated change, would be easy to put somewhere else.

File: plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu.hs [hunk 2]
```diff
@@ -1,9 +1,10 @@
+{-# LANGUAGE CPP                      #-}
```
Comment (parsonsmatt) [inline plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu.hs:1]: This is used to provide backwards compatibility.

File: plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu.hs [hunk 3]
```diff
@@ -1,9 +1,10 @@
...
+{-# LANGUAGE DataKinds                #-}
```
Comment (parsonsmatt) [inline plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu.hs:2]: This extension is reshuffled because I ran `stylish-haskell` (which appears to be how this file is formatted) and it automatically did that change.

Same for the reordered imports below.

File: plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu.hs [hunk 4]
```diff
@@ -23,16 +24,18 @@ import           Development.IDE.GHC.Compat      as Compat hiding (Cpp)
...
 import           Ide.Plugin.Properties
-import           Ide.PluginUtils                 (makeDiffTextEdit, usePropertyLsp)
+import           Ide.PluginUtils                 (makeDiffTextEdit,
+                                                  usePropertyLsp)
 import           Ide.Types
 import           Language.LSP.Server             hiding (defaultConfig)
 import           Language.LSP.Types
 import           Language.LSP.Types.Lens         (HasTabSize (tabSize))
 import           Ormolu
+import           Ormolu.Config
```
Comment (parsonsmatt) [inline plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu.hs:34]: This is redundant for `fourmolu <= 0.7`. I didn't gate it behind CPP because CPP is ugly.

File: plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal [hunk 5]
```diff
@@ -9,21 +9,27 @@ license:            Apache-2.0
...
+
 library
   exposed-modules:  Ide.Plugin.Fourmolu
   hs-source-dirs:   src
   ghc-options:      -Wall
   build-depends:
     , base            >=4.12 && <5
     , filepath
-    , fourmolu        ^>=0.3 || ^>=0.4 || ^>= 0.6
+    , fourmolu        ^>=0.3 || ^>=0.4 || ^>= 0.6 || ^>= 0.7
```
Comment (georgefst) [inline plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal:32]: This is in the wrong commit, right? Not necessarily actually worth changing now.

File: plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu.hs (hunk 3 above)
Comment (georgefst) [reply plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu.hs:2]: So `stylish-haskell` isn't enforced by CI? If not, I'd be tempted to go back to `fourmolu`.

This file was formatted with `stylish-haskell` when it looked like that was becoming necessary (#693, #1439). But I lost track of where the decision on formatting this codebase ended up.

File: plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal [hunk 6]
```diff
@@ -29,7 +29,8 @@ library
   build-depends:
     , base            >=4.12 && <5
     , filepath
-    , fourmolu        ^>=0.3 || ^>=0.4 || ^>= 0.6 || ^>= 0.7
+    -- , fourmolu        ^>=0.3 || ^>=0.4 || ^>= 0.6 || ^>= 0.7
```
Comment (georgefst) [inline plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal]: Eugh, I take it we still can't drop support for old versions due to `stylish-haskell`: https://github.com/haskell/haskell-language-server/pull/2254#discussion_r721196303?

(if we could, we could obviously remove all the CPP)

**Review by georgefst — APPROVED** [review]
Commit history is a bit weird, but this otherwise looks good!

File: plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal (hunk 5 above)
Comment (parsonsmatt) [reply plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal:32]: I am regrettably a complete `git` noob and generally just squash+merge all my stuff 😅

File: plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu.hs (hunk 3 above)
Comment (pepeiborra) [reply plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu.hs:2]: We have a pre-commit hook that runs stylish-haskell, mentioned in the contributing guidelines. But it is not enforced by CI:

https://github.com/haskell/haskell-language-server/blob/7431cea783f3b92d327635c017dd5de87b4b8774/docs/contributing/contributing.md#L171-L214

File: plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal (hunk 6 above)
Comment (pepeiborra) [reply plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal]: stylish-haskell has been updated since #2254  and I think it's very likely that this can be removed. Do you want to check in a follow-up PR?

**Comment by July541** [comment]
I checked the test detail and found all tests are run for fourmolu 0.3.0.0-0.6.0.0 :(

File: plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal (hunk 5 above)
Comment (georgefst) [reply plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal:32]: Ha, I'm very much in the "git is a wonderful concept with a horrible CLI" camp. Since I finally got around to setting up good editor integration, I haven't looked back.

File: plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal (hunk 6 above)
Comment (georgefst) [reply plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal]: See https://github.com/haskell/haskell-language-server/pull/2950.

**Comment by georgefst** [comment]
> I checked the test detail and found all tests are run for fourmolu 0.3.0.0-0.6.0.0 :(

Yep, https://github.com/haskell/haskell-language-server/pull/2950 fixes the test, and optimistically drops support for old versions (lets see if that passes CI...).
